packages/frontend/app/services/channels/find-or-create.ts

```diff
@@ -0,0 +1,128 @@
+import ContactManager from '../contact-manager';
+import Identity from 'emberclear/models/identity';
+import CurrentUserService from '../current-user';
+
+export default class FindOrCreateChannelService extends Service {
+  @service store!: StoreService;
+  @service contactManager!: ContactManager;
+  @service currentUser!: CurrentUserService;
+
+  async findOrCreateChannel(channelInfo: StandardMessage['channelInfo']): Promise<Channel> {
+    const { uid, name, members, admin, activeVotes, contextChain } = channelInfo!;
+
+    try {
+      return await this.store.findRecord('channel', uid);
+    } catch (e) {
+      const channel = this.store.createRecord('channel', {
+        uid,
+        name,
+        admin: await this.findOrCreateMember(admin),
+        members: members.map(async (member) => await this.findOrCreateMember(member)),
+        activeVotes: activeVotes.map(async (vote) => await this.findOrCreateVote(vote)),
+        contextChain: await this.findOrCreateContextChain(contextChain),
+      });
+
+      return channel;
+    }
+  }
+
+  async findOrCreateContextChain(
+    standardContextChain: StandardChannelContextChain
+  ): Promise<ChannelContextChain | undefined> {
+    if (standardContextChain === undefined) {
+      return undefined;
+    }
+
+    const { id } = standardContextChain;
+
+    let contextChain = undefined;
+
+    try {
+      contextChain = await this.store.findRecord('channelContextChain', id);
+      return contextChain;
+    } catch (e) {
+      contextChain = await this.store.createRecord('channelContextChain', {
+        id,
+        admin: await this.findOrCreateMember(standardContextChain.admin),
+        members: standardContextChain.members.map((member) => this.findOrCreateMember(member)),
+        voteChain: await this.findOrCreateVote(standardContextChain.supportingVote),
+        previousVoteChain: await this.findOrCreateContextChain(standardContextChain.previousChain),
+      });
+      return contextChain;
+    }
+  }
+
+  async findOrCreateVote(standardVote: StandardVote): Promise<Vote | undefined> {
+    if (standardVote === undefined) {
+      return undefined;
+    }
+
+    const { id } = standardVote;
+
+    let vote = undefined;
+
+    try {
+      vote = await this.store.findRecord('vote', id);
+      return vote;
+    } catch (e) {
+      vote = this.store.createRecord('vote', {
+        id,
+        voteChain: await this.findOrCreateVoteChain(standardVote.voteChain),
+      });
+      return vote;
+    }
+  }
+
+  async findOrCreateVoteChain(
+    standardVoteChain: StandardVoteChain
+  ): Promise<VoteChain | undefined> {
+    if (standardVoteChain === undefined) {
+      return undefined;
+    }
+
+    const { id } = standardVoteChain;
+
+    let voteChain = undefined;
+
+    try {
+      voteChain = await this.store.findRecord('voteChain', id);
+      return voteChain;
+    } catch (e) {
+      voteChain = this.store.createRecord('voteChain', {
```

Comment:
We forgot to put signature in StandardVoteChain

Reply:
We'll need to go through the process of converting back from hex string that we did whenever we were messing with the vote verifier stuff. I think we already have the helper methods for that, so it should be pretty straightforward.